### PR TITLE
ROCm 3.5 and 3.7 CMake fixes

### PR DESCRIPTION
**Pull request by mkuron**
HCC was removed and replaced by HIP-Clang. The CMake linker helper was evidently not updated, but can be circumvented. Since HCC is gone now, we hopefully won't have to modify the CMake integration for every update anymore.

@biermanncarl @jngrad

### CODE REVIEW

**Review by biermanncarl — DISMISSED** [review]

**Comment by jngrad** [comment]
@mkuron A few GPU electrostatics tests are timing out on `lama`:
```
115/148 Test  #25: coulomb_cloud_wall ..........................***Timeout 300 sec
116/148 Test  #26: coulomb_tuning ..............................***Timeout 300 sec
118/148 Test  #29: dawaanr-and-bh-gpu ..........................***Timeout 300 sec
130/148 Test  #86: coulomb_cloud_wall_duplicated ...............***Timeout 300 sec
```

File: cmake/FindCUDACompilerHIP.cmake [hunk 1]
```diff
@@ -27,15 +27,18 @@ find_package(HIP ${CUDACompilerHIP_FIND_VERSION} MODULE REQUIRED)
...
+  set(HIP_HIPCC_CMAKE_LINKER_HELPER "${HIP_HIPCC_EXECUTABLE}")
+  unset(HCC_PATH)
 endif()
 find_package(HIP ${CUDACompilerHIP_FIND_VERSION} MODULE REQUIRED)
 
 set(CUDA 1)
 set(HIP 1)
 
-list(APPEND HIP_HCC_FLAGS
+list(APPEND HIP_HIPCC_FLAGS
```
Comment (jngrad) [inline cmake/FindCUDACompilerHIP.cmake:41]: is this variable name change compatible with versions before 3.5?

**Comment by jngrad** [comment]
Do the GPU flags `HIP_HIPCC_FLAGS_*` also need to change name?

**Comment by mkuron** [comment]
HIP_HCC_FLAGS is specific to HCC and HIP_HIPCC_FLAGS is the "generic" one passed to whatever compiler backend the hipcc wrapper script is using. This change is backwards compatible, see https://github.com/ROCm-Developer-Tools/HIP/blob/master/cmake/FindHIP.cmake.

**Comment by mkuron** [comment]
> A few GPU electrostatics tests are timing out on lama

That's odd, they get stuck here:
```
#0  0x000015555503ee57 in sched_yield () at ../sysdeps/unix/syscall-template.S:78
#1  0x000015552f703b55 in amd::AqlQueue::ExecutePM4(unsigned int*, unsigned long) () from /opt/rocm/lib/libhsa-runtime64.so.1
#2  0x000015552f6f9798 in amd::GpuAgent::InvalidateCodeCaches() () from /opt/rocm/lib/libhsa-runtime64.so.1
#3  0x000015552f705fae in amd::LoaderContext::SegmentAlloc(amdgpu_hsa_elf_segment_t, hsa_agent_s, unsigned long, unsigned long, bool) () from /opt/rocm/lib/libhsa-runtime64.so.1
#4  0x000015552f73fc85 in amd::hsa::loader::ExecutableImpl::LoadSegmentsV2(hsa_agent_s, amd::hsa::code::AmdHsaCode const*) () from /opt/rocm/lib/libhsa-runtime64.so.1
#5  0x000015552f74037e in amd::hsa::loader::ExecutableImpl::LoadSegments(hsa_agent_s, amd::hsa::code::AmdHsaCode const*, unsigned int) () from /opt/rocm/lib/libhsa-runtime64.so.1
#6  0x000015552f743463 in amd::hsa::loader::ExecutableImpl::LoadCodeObject(hsa_agent_s, hsa_code_object_s, unsigned long, char const*, hsa_loaded_code_object_s*) ()
   from /opt/rocm/lib/libhsa-runtime64.so.1
#7  0x000015552f7182d7 in HSA::hsa_executable_load_agent_code_object(hsa_executable_s, hsa_agent_s, hsa_code_object_reader_s, char const*, hsa_loaded_code_object_s*) ()
   from /opt/rocm/lib/libhsa-runtime64.so.1
#8  0x000015554ef9a925 in roc::LightningProgram::setKernels(amd::option::Options*, void*, unsigned long) () from /opt/rocm/lib/libamdhip64.so.3
#9  0x000015554ef43f15 in device::Program::linkImplLC(amd::option::Options*) () from /opt/rocm/lib/libamdhip64.so.3
#10 0x000015554ef44b45 in device::Program::build(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const*, amd::option::Options*) ()
   from /opt/rocm/lib/libamdhip64.so.3
#11 0x000015554ef5778c in amd::Program::build(std::vector<amd::Device*, std::allocator<amd::Device*> > const&, char const*, void (*)(_cl_program*, void*), void*, bool) ()
   from /opt/rocm/lib/libamdhip64.so.3
#12 0x000015554eee2ea5 in PlatformState::getFunc(void const*, int) () from /opt/rocm/lib/libamdhip64.so.3
#13 0x000015554eeeb6e3 in hipLaunchKernel () from /opt/rocm/lib/libamdhip64.so.3
#14 0x0000155551a1b839 in void thrust::hip_rocprim::parallel_for<thrust::hip_rocprim::tag, thrust::hip_rocprim::__uninitialized_fill::functor<thrust::device_ptr<CUDA_particle_data>, CUDA_particle_data>, unsigned long>(thrust::hip_rocprim::execution_policy<thrust::hip_rocprim::tag>&, thrust::hip_rocprim::__uninitialized_fill::functor<thrust::device_ptr<CUDA_particle_data>, CUDA_particle_data>, unsigned long)::workaround::par(thrust::hip_rocprim::execution_policy<thrust::hip_rocprim::tag>&, thrust::hip_rocprim::__uninitialized_fill::functor<thrust::device_ptr<CUDA_particle_data>, CUDA_particle_data>, unsigned long) () from /tikhome/mkuron/Documents/espresso/build/src/core/EspressoCore.so
#15 0x0000155551a1b641 in thrust::detail::vector_base<CUDA_particle_data, CudaDeviceAllocator<CUDA_particle_data> >::vector_base(unsigned long) ()
   from /tikhome/mkuron/Documents/espresso/build/src/core/EspressoCore.so
#16 0x0000155551a1b2aa in void resize_or_replace<CUDA_particle_data>(thrust::device_vector<CUDA_particle_data, CudaDeviceAllocator<CUDA_particle_data> >&, unsigned long) ()
   from /tikhome/mkuron/Documents/espresso/build/src/core/EspressoCore.so
#17 0x0000155551a1a374 in resize_buffers(unsigned long) () from /tikhome/mkuron/Documents/espresso/build/src/core/EspressoCore.so
#18 0x0000155551a1a8ef in copy_part_data_to_gpu(ParticleRange) () from /tikhome/mkuron/Documents/espresso/build/src/core/EspressoCore.so
#19 0x00001555518d398d in EspressoSystemInterface::gatherParticles() () from /tikhome/mkuron/Documents/espresso/build/src/core/EspressoCore.so
#20 0x00001555518d7965 in force_calc(CellStructure&) () from /tikhome/mkuron/Documents/espresso/build/src/core/EspressoCore.so
#21 0x00001555518f422c in integrate(int, int) () from /tikhome/mkuron/Documents/espresso/build/src/core/EspressoCore.so
#22 0x0000155551899bc7 in mpi_integrate(int, int) () from /tikhome/mkuron/Documents/espresso/build/src/core/EspressoCore.so
```
Also seeing this in dmesg:
```
[ 9422.510382] amdgpu 0000:43:00.0: [gfxhub0] no-retry page fault (src_id:0 ring:40 vmid:8 pasid:32774, for process python3 pid 108190 thread python3 pid 108190)
[ 9422.510385] amdgpu 0000:43:00.0:   in page starting at address 0x000000000002d000 from client 27
[ 9422.510386] amdgpu 0000:43:00.0: VM_L2_PROTECTION_FAULT_STATUS:0x00841051
[ 9422.510387] amdgpu 0000:43:00.0: 	 Faulty UTCL2 client ID: 0x8
[ 9422.510388] amdgpu 0000:43:00.0: 	 MORE_FAULTS: 0x1
[ 9422.510389] amdgpu 0000:43:00.0: 	 WALKER_ERROR: 0x0
[ 9422.510390] amdgpu 0000:43:00.0: 	 PERMISSION_FAULTS: 0x5
[ 9422.510391] amdgpu 0000:43:00.0: 	 MAPPING_ERROR: 0x0
[ 9422.510392] amdgpu 0000:43:00.0: 	 RW: 0x1
[ 9422.510402] amdgpu 0000:43:00.0: [gfxhub0] no-retry page fault (src_id:0 ring:40 vmid:8 pasid:32774, for process python3 pid 108190 thread python3 pid 108190)
[ 9422.510403] amdgpu 0000:43:00.0:   in page starting at address 0x000000000007c000 from client 27
[ 9422.510404] amdgpu 0000:43:00.0: VM_L2_PROTECTION_FAULT_STATUS:0x00000000
[ 9422.510405] amdgpu 0000:43:00.0: 	 Faulty UTCL2 client ID: 0x0
[ 9422.510405] amdgpu 0000:43:00.0: 	 MORE_FAULTS: 0x0
[ 9422.510406] amdgpu 0000:43:00.0: 	 WALKER_ERROR: 0x0
[ 9422.510407] amdgpu 0000:43:00.0: 	 PERMISSION_FAULTS: 0x0
[ 9422.510408] amdgpu 0000:43:00.0: 	 MAPPING_ERROR: 0x0
[ 9422.510409] amdgpu 0000:43:00.0: 	 RW: 0x0
```

I don't have time to take care of that right now, but will do so before we upgrade the Docker containers to 3.5. This pull request can be merged anyway as it's pure CMake and is needed by @biermanncarl for local testing.

**Comment by jngrad** [comment]
We can't merge if we know it's broken. @biermanncarl can do the following in his PR:
```sh
git fetch upstream pull/3746/head:pr-3746
git cherry-pick db92aead..pr-3746
```
Once you've fixed this PR, he can revert those changes and merge `python`.

**Comment by mkuron** [comment]
> We can't merge if we know it's broken.

This pull request is broken in the same way as the current master. It does not change behavior for ROCm < 3.5. By your reasoning, we could never merge any pull requests as we always know that there unfixed bugs in Espresso.

**Comment by jngrad** [comment]
> By your reasoning, we could never merge any pull requests as we always know that there unfixed bugs in Espresso.

That's not my reasoning, and you know that. This PR introduces changes that are untested for the 3.5 case in CI and that either break for >=3.5 or change nothing for <3.5. Merging 2 PRs is only going to make backporting in 4.1.3 more difficult for me.

Your commits can be temporarily merged or cherry-picked in the SD PR because it's a WIP that will be rebased eventually.

**Comment by mkuron** [comment]
ROCm 3.7 support added. All tests complete successfully, @jngrad.

**Review by jngrad — APPROVED** [review]

**Comment by mkuron** [comment]
Unrelated test failure on the Mac CI job. Please click the retry button, @jngrad.

**Pull request merged**